SNOW-618683 Add previous status code as parameter "retry", update tests

sfc-gh-mknister opened the pull request:
# Overview

SNOW-618683

I have read the CLA Document and I hereby sign the CLA

If the driver issues a retry, the query parameter "retry" is now added. "Retry" contains the value of the HTTP status code from the server's previous response. If the response or the status code is null, then the value is set as "0" (as a string value). The parameter "retryCount" is already added.

Note that neither "retry" nor "retryCount" parameters will be included when:
1.) The driver is fetching using a AWS presigned URL because this causes a bug
2.) The driver is issuing a login request. I am not sure if there's any real reason not to include these query parameters for the login request so they could be added back in if needed.

Link to dummy PR (merge to master) so that precommit jobs run: https://github.com/snowflakedb/snowflake-jdbc/pull/1225

## External contributors - please answer these questions before submitting a pull request. Thanks!

Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN 


2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

## Pre-review checklist
- [ ] This change has passed precommit
- [ ] I have reviewed code coverage report for my PR in  ([Sonarqube](https://sonarqube.int.snowflakecomputing.com/project/branches?id=snowflake-jdbc))

sfc-gh-mknister commented on src/main/java/net/snowflake/client/jdbc/RestRequest.java:
```diff
@@ -346,6 +349,8 @@ public static CloseableHttpResponse execute(
         }
 
         retryCount++;
+        lastStatusCodeForRetry =
```
When there is no response or status code, the value of `retry` will be the string value "null". However, we should never hit this case because the JDBC driver does not retry if the response is null.

sfc-gh-hchaturvedi replied:
@sfc-gh-mknister have you tested the case where there is no response from the backend resulting in a timeout? What is the error code in that case? Also instead of using null can we choose a number that is less risky? Like 0? Is 0 already used for something?

sfc-gh-mknister replied:
@sfc-gh-hchaturvedi if there is no response from the backend, the response will come back as null. I tested this manually by using Burpsuite and intercepting the POST requests, and then waiting until the driver timed out. In this case, it is possible to do a retry.

Looking at the list of [status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status), it looks like 0 is not used for anything! Changed the code to `0` for the null case instead.

sfc-gh-mknister commented on src/test/java/net/snowflake/client/jdbc/ConnectionIT.java:
```diff
@@ -62,39 +61,6 @@ public void testSimpleConnection() throws SQLException {
     con.close(); // ensure no exception
   }
 
-  @Test
```
Please ignore- removing unneccessary test that got merged by accident and is causing failures.

sfc-gh-mknister commented:
I have read the CLA Document and I hereby sign the CLA

sfc-gh-hchaturvedi commented:
@sfc-gh-mknister Can we add a test for timeout connection? Also can you think of a way to prevent future developers from calling this API with` includeRetryPArameters = true` unless it is to execute a statement?

sfc-gh-twang commented on src/main/java/net/snowflake/client/jdbc/RestRequest.java:
```diff
@@ -346,6 +349,8 @@ public static CloseableHttpResponse execute(
 
         retryCount++;
+        lastStatusCodeForRetry =
+            response == null ? "0" : String.valueOf(response.getStatusLine().getStatusCode());
```
Thanks for making the change.
A few questions:
1. Is it intended to add "retry" query parameter on all requests from JDBC? 
    I think the asks was to (a) make sure all connectors sets the existing "retry" query parameter correctly when hitting /queries/v1/query-request endpoint, and (b) add a "retryReason" query parameter when hitting the same endpoint. 
    It looks like this PR only modifies JDBC and it sets "retry" query parameter but not "retryReason" parameter, which I think is a bit different from the original ask.
2. According to source code of GS's [/queries/v1/query-request endpoint](https://sourcegraph.c2.us-central1.gcp-dev.app.snowflake.com/github.com/snowflakedb/snowflake/-/blob/GlobalServices/src/main/java/com/snowflake/resources/QueriesResource.java?L666:19), the endpoint expects the value of "retry" parameter to be "true" or "false". However, this PR sets the value of "retry" parameter to the last request's status code. Even if we update the logic on the GS side, I think we will have backward compatibility issue, unless we add a new  /queries/**v2**/query-request endpoint, which I don't think is necessary for our purpose.

sfc-gh-mknister replied:
@sfc-gh-twang this has been updated to rename the query parameter to "retryReason" instead of just "retry"

sfc-gh-twang replied:
@sfc-gh-mknister 
Thanks for adding retryReason query parameter!
A few questions:
The PR only modifies JDBC, right? 
Does JDBC now set the retry  and retryCount query parameters?

I think the original asks were to 
(a) make sure all connectors sets the existing "retry" and "retryCount" query parameters correctly when hitting /queries/v1/query-request endpoint, and 
(b) add a "retryReason" query parameter when hitting the same endpoint.

I think your PR did (b), but not sure about (a).
And I was wondering what about other connectors, are they going to come in other PRs and this is just the first PR?

sfc-gh-mknister replied:
@sfc-gh-twang to answer your questions:

This PR only modifies JDBC. Each connector is in a different repository so each driver will have a separate PR for this task.

The JDBC already sets "retryCount" when issuing a retry. It does not and has never set the boolean query parameter "retry". Looking at the GS code logic, I don't think that "retry" needs to be set to true as long as "retryCount" is incremented. The only place "retry" is used is in this line, and it's in an OR clause with "retryCount > 0". https://github.com/snowflakedb/snowflake/blob/main/GlobalServices/src/main/java/com/snowflake/resources/QueriesResource.java#L707 Is there a reason that both should be set?

This PR adds the query parameter retryReason, which will provide an integer value of the last HTTP status code returned from GS from the previous response. If the retry is due to a socket timeout and there is no status code available, the retryReason will be set to 0.

sfc-gh-twang reviewed (changes requested):
I'm not familiar with JDBC code, but I left some questions about the new behavior.

sfc-gh-mknister commented:
> @sfc-gh-mknister Can we add a test for timeout connection? Also can you think of a way to prevent future developers from calling this API with` includeRetryPArameters = true` unless it is to execute a statement?

@sfc-gh-hchaturvedi a test has been added for the socketTimeoutException case. I also added a comment about the "includeRetryParameters" parameter in RestRequest.java. I don't think there's any real harm in developers setting this parameter to true if they'd want those query params. We added this param and just need to ensure it's set to false for fetching results from the S3 bucket.

sfc-gh-twang reviewed (approved):
LGTM! Left a few more high-level questions for the overall effort and plan.
